cpp/src/arrow/util/bit_block_counter.h

```diff
@@ -48,43 +48,33 @@ inline uint64_t ShiftWord(uint64_t current, uint64_t next, int64_t shift) {
 // These templates are here to help with unit tests
 
 template <typename T>
-struct BitBlockAnd {
-  static T Call(T left, T right) { return left & right; }
-};
+inline T BitNot(T x) {
+  return ~x;
+}
 
 template <>
-struct BitBlockAnd<bool> {
-  static bool Call(bool left, bool right) { return left && right; }
-};
+inline bool BitNot(bool x) {
+  return !x;
+}
 
 template <typename T>
-struct BitBlockAndNot {
-  static T Call(T left, T right) { return left & ~right; }
+struct BitBlockAnd {
```

Comment:
Since we're using a plain function for `BitNot`, can we do the same for this one too (and the other functors)?

Reply:
Can't change the `BitBlockXXX` into functions because these are passed [template parameters (Operators)](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bit_block_counter.h#L289) and the [templetized function invokes `Op::Call(...)`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bit_block_counter.h#L308).

Reply:
`BitNot` was not an existing operator, I added it because negation is the only bitwise operation where C++ behaves different between bool and integral types, so I specialized it as a helper function.
`~(bool)x` always returns true.

Reply:
I see, thanks.